sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -119,23 +146,139 @@ static void *_event_loop(void *arg)
+        } else if (prev_state == NO_SPACE) {
+            /* No space in session management. Should not happen. If it occurs,
+            we lost track of sessions */
+            DEBUG("gcoap: no space in session management. We lost track of sessions!")
+            sock_dtls_session_destroy(sock, &socket.ctx_dtls_session);
+        }
+
+        /* If not enough session slots left: set timeout to free up session */
+        uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+        if (dsm_get_num_available_slots() < minimum_free)
+        {
+            uint32_t timeout = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS_TIMEOUT_USEC;
+            event_callback_init(&_dtls_session_free_up_tmout_cb,
+                                _dtls_free_up_session, NULL);
+            event_timeout_init(&_dtls_session_free_up_tmout, &_queue,
+                               &_dtls_session_free_up_tmout_cb.super);
+            event_timeout_set(&_dtls_session_free_up_tmout, timeout);
+        }
+    }
+
+    if (type & SOCK_ASYNC_CONN_FIN) {
+        if (sock_dtls_get_event_session(sock, &socket.ctx_dtls_session)) {
+            /* Session is already destroyed, only remove it from dsm */
+            dsm_remove(sock, &socket.ctx_dtls_session);
+        } else {
+            puts("gcoap: A session was closed, but the corresponding session " \
+            "could not be retrieved from the socket!");
+            return;
+        }
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+
+        /* Remove all memos of the concerned session. TODO: oberservable memos! */
+        for (int i = 0; i < CONFIG_GCOAP_REQ_WAITING_MAX; i++) {
+            if (_coap_state.open_reqs[i].state == GCOAP_MEMO_UNUSED) {
+                continue;
+            }
+            gcoap_request_memo_t *memo = &_coap_state.open_reqs[i];
+            if (sock_udp_ep_equal(&memo->remote_ep, &ep)) {
+                _expire_request(memo);
+                event_timeout_clear(&memo->resp_evt_tmout);
+            }
+        }
+    }
+
+    if (type & SOCK_ASYNC_MSG_RECV) {
+        ssize_t res = sock_dtls_recv(sock, &socket.ctx_dtls_session, _listen_buf,
+                                    sizeof(_listen_buf), 0);
+        if (res <= 0) {
+            DEBUG("gcoap: DTLS recv failure: %d\n", (int)res);
+            return;
+        }
+        sock_udp_ep_t ep;
+        sock_dtls_session_get_udp_ep(&socket.ctx_dtls_session, &ep);
+        _process_coap_pdu(&socket, &ep,  _listen_buf, res);
+    }
+}
+
+/* Timeout function to free up a session when too many session slots are occupied */
+static void _dtls_free_up_session(void *arg) {
+    (void)arg;
+    sock_dtls_session_t session;
+
+    uint8_t minimum_free = CONFIG_GCOAP_DTLS_MINIMUM_AVAILABLE_SESSIONS;
+    if (dsm_get_num_available_slots() < minimum_free) {
+        if(dsm_get_least_recently_used_session(&_sock_dtls, &session) != -1) {
```

Comment:
> I think there is still something wrong here:
> When playing around with two native RIOT nodes querying a third native RIOT node with
> 
> ```
> coap get fe80::7837:fcff:fe7d:1aaf 5684 /.well-known/core
> ```
> 
> after a while one of the nodes would no longer receive a response. (This is after the `dsm: removed session` message), the server shows
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 1/1
>  CLI requests sent: 0
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The clients meanwhile display
> 
> ```
> CoAP server is listening on port 5684
> Connection secured with DTLS
> Free DTLS session slots: 13/1
>  CLI requests sent: 17
> CoAP open requests: 0
> Configured Proxy: None
> ```
> 
> The number of 'free sessions' increases with each request that does not receive a response.

Fixed. It was a problem with finding sessions which have been removed but stated them as "already existing", since session state was not taken into account.